### DISCUSSION TITLE
Dark Background Color of the Conversion Window in PC

### DISCUSSION THREAD

**Discussion by sk-akiyama** [google/mozc#940] [Ideas]
![image](https://github.com/google/mozc/assets/48663185/d0f0d861-e3f0-439a-8cd0-b05e5c598a23)

Please implement the option to enable dark mode for the background color of conversion window in pc.
Many people are using dark mode on their systems lately, and I think it would be much easier on the eyes.

**Comment by yukawa** [comment]
We've been long aware of this feature request, but there remain several other super high priority projects that we want to focus our resources on first (e.g. build system migration from deprecated [GYP](https://gyp.gsrc.io) to [Bazel](https://bazel.build)) because otherwise it would become much more difficult for us to keep maintaining this project. So, yes, we will be working on this feature in the future, but not now. See also https://github.com/google/mozc/discussions/796#discussioncomment-6808915 and https://github.com/google/mozc/discussions/864#discussioncomment-8204518.

In the meantime, if you want to modify the color scheme and build your own customized version of Mozc, you can start from looking into the following code locations.

https://github.com/google/mozc/blob/4587e5071d05cb293a4c1b0cbc4c3ff395df5a10/src/renderer/renderer_style_handler.cc#L77-L189
https://github.com/google/mozc/blob/4587e5071d05cb293a4c1b0cbc4c3ff395df5a10/src/renderer/win32/text_renderer.cc#L70-L112

**Reply by sk-akiyama** [reply]
Oh! Thanks! I'll try it!

**Reply by sk-akiyama** [reply]
@yukawa 

Hello!
Thank you for answering my questions last time.

I tried the following build command based on [this page](https://github.com/google/mozc/blob/master/docs/build_mozc_in_windows.md#build-mozc), but encountered an error. Could you please provide assistance on how to resolve it?

```python build_mozc.py build -c Release package```

Error Message:
```
FAILED: protoc.exe protoc.exe.pdb
C:\Users\Kousuke\AppData\Local\Programs\Python\Python312\python.exe gyp-win-tool link-with-manifests environment.x86 True protoc.exe "C:\Users\Kousuke\AppData\Local\Programs\Python\Python312\python.exe gyp-win-tool link-wrapper environment.x86 False link.exe /nologo /OUT:protoc.exe @protoc.exe.rsp" 1 mt.exe rc.exe "obj\protobuf\protoc.protoc.exe.intermediate.manifest" obj\protobuf\protoc.protoc.exe.generated.manifest
Traceback (most recent call last):
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 332, in <module>
    sys.exit(main(sys.argv[1:]))
             ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 31, in main
    exit_code = executor.Dispatch(args)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 73, in Dispatch
    return getattr(self, method)(*args[1:])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 136, in ExecLinkWrapper
    out, _ = link.communicate()
             ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Kousuke\AppData\Local\Programs\Python\Python312\Lib\subprocess.py", line 1196, in communicate
    stdout = self.stdout.read()
             ^^^^^^^^^^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x83 in position 0: invalid start byte
Traceback (most recent call last):
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 332, in <module>
    sys.exit(main(sys.argv[1:]))
             ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 31, in main
    exit_code = executor.Dispatch(args)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 73, in Dispatch
    return getattr(self, method)(*args[1:])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Kousuke\Documents\mozc\src\out_win\Release\gyp-win-tool", line 182, in ExecLinkWithManifests
    subprocess.check_call(ldcmd + add_to_ld)
  File "C:\Users\Kousuke\AppData\Local\Programs\Python\Python312\Lib\subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'C:\Users\Kousuke\AppData\Local\Programs\Python\Python312\python.exe gyp-win-tool link-wrapper environment.x86 False link.exe /nologo /OUT:protoc.exe @protoc.exe.rsp protoc.exe.manifest.res' returned non-zero exit status 1.
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "C:\Users\Kousuke\Documents\mozc\src\build_mozc.py", line 898, in <module>
    sys.exit(main())
             ^^^^^^
  File "C:\Users\Kousuke\Documents\mozc\src\build_mozc.py", line 883, in main
    BuildMain(cmd_opts, cmd_args)
  File "C:\Users\Kousuke\Documents\mozc\src\build_mozc.py", line 651, in BuildMain
    BuildOnWindows(targets)
  File "C:\Users\Kousuke\Documents\mozc\src\build_mozc.py", line 633, in BuildOnWindows
    RunOrDie([ninja, '-C', build_arg, target_name])
  File "C:\Users\Kousuke\Documents\mozc\src\build_tools\util.py", line 140, in RunOrDie
    raise RunOrDieError('\n'.join(['',
build_tools.util.RunOrDieError:
==========
 ERROR: C:\Users\Kousuke\Documents\mozc\src\third_party\ninja\ninja.exe -C out_win/Release mozc_win32_build32
==========

```

I believe it's likely related to encoding, but I'm not sure...

Thank you.

**Reply by yukawa** [reply]
> I tried the following build command based on [this page](https://github.com/google/mozc/blob/master/docs/build_mozc_in_windows.md#build-mozc), but encountered an error. Could you please provide assistance on how to resolve it?

Sorry, I have no idea too. In the meantime, I'd suggest using GitHub Actions to build Mozc as explained below.
https://github.com/google/mozc/blob/master/docs/build_mozc_in_windows.md#build-with-github-actions

**Reply by sk-akiyama** [reply]
OK! Thanks!!

**Reply by sk-akiyama** [reply]
@yukawa

Hi!
I was able to create an .msi file and install it since then!
And I was also able to change the color of the text. (I've set it to gray in the image.)
However, I still can't seem to change the background color...

![2024-06-04_15h05_18](https://github.com/google/mozc/assets/48663185/658ffe25-d948-4c0f-a0e7-4c1279262414)

Here's the forked repository:
https://github.com/sk-akiyama/mozc/commits/master/

Thank you for your help in finding a solution.

**Reply by yukawa** [reply]
> However, I still can't seem to change the background color...

Perhaps you may need to edit the following lines to change the background color of the candidate window.
https://github.com/google/mozc/blob/864c5433259188015d0b5b1a09f8cc4f87bd8907/src/renderer/win32/candidate_window.cc#L82-L91

Also for `indicator_window`.
https://github.com/google/mozc/blob/864c5433259188015d0b5b1a09f8cc4f87bd8907/src/renderer/win32/indicator_window.cc#L252-L304

**Reply by sk-akiyama** [reply]
@yukawa 
Thank you!
I really appreciate it!!

**Comment by sk-akiyama** [comment]
@yukawa 
Hi, Long time no see!
Thanks to your previous guidance, I was able to enable dark mode support on Windows. However, I can’t seem to achieve the same on macOS.

Could you please let me know where I can modify the background color to support dark mode on macOS?

Thank you for your time and support!

<img width="214" alt="2024-12-15_12-35-58" src="https://github.com/user-attachments/assets/ee69e9ab-7c50-4333-bd18-9ff2d6aa04ed" />

**Reply by yukawa** [reply]
Hmm, I'm not familiar with `mozc_renderer` for macOS.

What I can easily point out is that `renderer_style.proto` doesn't look to define a background color for the window. So you probably need to directly set the background color in macOS-specific window management code somewhere in [src/renderer/mac/](https://github.com/google/mozc/tree/master/src/renderer/mac). Note that [src/renderer/win32/](https://github.com/google/mozc/tree/master/src/renderer/win32) is for Windows-specific code. Changing files there should not have any effect for macOS build.

**Reply by sk-akiyama** [reply]
@yukawa 
Oh… I see…
I looked through the mac/ folder, but I still can’t figure out how to set the background color… (crying)
Do you have any tips or suggestions for solving this? I feel pretty lost here…

**Reply by yukawa** [reply]
> I looked through the mac/ folder, but I still can’t figure out how to set the background color… (crying)
> Do you have any tips or suggestions for solving this? I feel pretty lost here…

I haven't confirmed it, but maybe this is the place where the window background color is specified.
https://github.com/google/mozc/blob/bc546b239bccff494773104e108ba08fcff40d5b/src/renderer/mac/RendererBaseWindow.mm#L65

**Comment by sk-akiyama** [comment]
@yukawa 
Thank you for the advice!

I tried it, but it didn’t seem to change anything…
Is it difficult to pinpoint the exact location?😢

**Reply by yukawa** [reply]
Sorry, I'm currently traveling and don't have access to any macOS environment until Jan 7, 2025.

**Reply by sk-akiyama** [reply]
Thanks! Have a good travel!
